### PR TITLE
Type `useApi` parameters improve docs

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

- type the fetcher
- type the errors
- pass the type parameters to `useSWR`

Note that the response type is still asserted, no actual parsing happens.

## Why?

Helps get better `fallbackData`, discovered in #11869

### CODE REVIEW

File: dotcom-rendering/src/lib/useApi.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	url: string = '',
+	options?: SWRConfiguration<T, E>,
+	init?: RequestInit,
+): ApiResponse<T, E> => {
+	const { data, error } = useSWR<T, E>(url, fetcher(init), options);
+
+	return {
+		data,
+		error,
+		loading: !url && isUndefined(error) && isUndefined(data),
```
Comment (JamieB-gu) [inline dotcom-rendering/src/lib/useApi.ts:49]: This has gone from `!!url` to `!url`? We're flipping the `boolean`?

File: dotcom-rendering/src/lib/useApi.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @template T assert the expected response type
+ * @template E assert the potential error type
+ * @param {string} url - The API endpoint. Falsy values will prevent any network requests
+ * @param {SWRConfiguration<T,E>} options - The SWR config object - https://swr.vercel.app/docs/api#options
+ * @param {RequestInit} init - The fetch init object - https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#supplying_request_options
+ * @returns {ApiResponse<T, E>}
+ * */
+export const useApi = <T = unknown, E = Error>(
+	url: string = '',
```
Comment (JamieB-gu) [inline dotcom-rendering/src/lib/useApi.ts:40]: Under what circumstances would you want to call this without a `url`?

File: dotcom-rendering/src/lib/useApi.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * A custom hook to make a GET request using the given url using the SWR lib (https://swr.vercel.app/).
+ *
+ * @template T assert the expected response type
+ * @template E assert the potential error type
+ * @param {string} url - The API endpoint. Falsy values will prevent any network requests
+ * @param {SWRConfiguration<T,E>} options - The SWR config object - https://swr.vercel.app/docs/api#options
+ * @param {RequestInit} init - The fetch init object - https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#supplying_request_options
+ * @returns {ApiResponse<T, E>}
+ * */
+export const useApi = <T = unknown, E = Error>(
```
Comment (JamieB-gu) [inline dotcom-rendering/src/lib/useApi.ts:39]: Could these type parameters have more descriptive names? E.g. `Res`, `Err`?

**Review by JamieB-gu — COMMENTED** [review]
Looks good, a few questions/suggestions.

File: dotcom-rendering/src/lib/useApi.ts (hunk 1 above)
Comment (mxdvl) [reply dotcom-rendering/src/lib/useApi.ts:49]: Whoops, an actual mistake, thanks for the spot:

```suggestion
		loading: !!url && isUndefined(error) && isUndefined(data),
```

File: dotcom-rendering/src/lib/useApi.ts (hunk 2 above)
Comment (mxdvl) [reply dotcom-rendering/src/lib/useApi.ts:40]: Ideally we would not… but I did not want to change the signature in fear of this breaking existing usages.

**Review by JamieB-gu — APPROVED** [review]

**Pull request merged**